File: loadercomplex-forge/ver18/src/main/java/com/enderzombi102/loadercomplex/forge18/impl/entity/ForgePlayer.java
package com.enderzombi102.loadercomplex.forge18.impl.entity;

import com.enderzombi102.loadercomplex.api.minecraft.entity.Player;
import com.enderzombi102.loadercomplex.api.minecraft.util.Gamemode;
import com.enderzombi102.loadercomplex.api.minecraft.util.Position;
import com.enderzombi102.loadercomplex.forge18.impl.utils.BlockUtils;
import net.minecraft.client.MinecraftClient;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.LiteralText;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.GameMode;

public class ForgePlayer extends ForgeLivingEntity implements Player {
	private final PlayerEntity wrappedEntity;

	public ForgePlayer(PlayerEntity backingEntity) {
		super(backingEntity);
		this.wrappedEntity = backingEntity;
	}

	@Override
	public boolean isSleeping() {
		return this.wrappedEntity.isSleeping();
	}

	@Override
	public Position getBedLocation() {
		return BlockUtils.toPosition( this.wrappedEntity.getSleepingPosition().orElseThrow() );
	}

	@Override
	public int getScore() {
		return this.wrappedEntity.getScore();
	}

	@Override
	public void addScore(int score) {
		this.wrappedEntity.addScore( score );
	}

	@Override
	public void sendMessage(String msg) {
		this.wrappedEntity.sendMessage( new LiteralText( msg ), false );
	}

	@Override
	public void setRespawnPoint(Position pos) {
		this.wrappedEntity.setSleepingPosition( new BlockPos( pos.x, pos.y, pos.z ) );
	}

	@Override
	public int getFoodLevel() {
		return this.wrappedEntity.getHungerManager().getFoodLevel();
	}

	@Override
	public void setFoodLevel(int level) {
		this.wrappedEntity.getHungerManager().setFoodLevel( level );
	}

	@Override
	public float getSaturationLevel() {
		return this.wrappedEntity.getHungerManager().getSaturationLevel();
	}

	@Override
	public void setSaturationLevel(float level) {
		this.wrappedEntity.getHungerManager().setSaturationLevel( level );
	}

	@Override
	public Gamemode getGamemode() {
		GameMode mode;
		if ( this.wrappedEntity.getWorld().isClient )
			//noinspection ConstantConditions
			mode = MinecraftClient.getInstance().interactionManager.getCurrentGameMode();
		else
			mode = ( (ServerPlayerEntity) this.wrappedEntity ).interactionManager.getGameMode();
		if ( mode == GameMode.DEFAULT )
			return Gamemode.SURVIVAL;
		return Gamemode.valueOf( mode.getName().toUpperCase() );
	}

	@Override
	public int getExperienceLevel() {
		return this.wrappedEntity.experienceLevel;
	}
}

File: loadercomplex-forge/src/ver182/kotlin/com/enderzombi102/loadercomplex/forge182/impl/entity/ForgePlayer.kt
package com.enderzombi102.loadercomplex.forge182.impl.entity

import com.enderzombi102.loadercomplex.api.math.Vec3i
import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity
import com.enderzombi102.loadercomplex.api.minecraft.util.Gamemode
import com.enderzombi102.loadercomplex.forge182.impl.utils.toLC
import net.minecraft.client.MinecraftClient
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.LiteralText
import net.minecraft.util.math.BlockPos
import net.minecraft.world.GameMode
import java.util.*
import net.minecraft.entity.player.PlayerEntity as McPlayerEntity

class ForgePlayer(private val wrappedEntity: McPlayerEntity) : ForgeLivingEntity( wrappedEntity ), PlayerEntity {
	override fun isSleeping(): Boolean =
		wrappedEntity.isSleeping

	override fun getBedLocation(): Optional<Vec3i> =
		wrappedEntity.sleepingPosition.map(BlockPos::toLC)

	override fun getScore(): Int =
		wrappedEntity.score

	override fun addScore(score: Int) {
		wrappedEntity.addScore(score)
	}

	override fun sendMessage(msg: String?) {
		wrappedEntity.sendMessage(LiteralText(msg), false)
	}

	override fun setRespawnPoint(pos: Vec3i) {
		wrappedEntity.setSleepingPosition(BlockPos(pos.x, pos.y, pos.z))
	}

	override fun getFoodLevel(): Int =
		wrappedEntity.hungerManager.foodLevel

	override fun setFoodLevel( level: Int ) {
		wrappedEntity.hungerManager.foodLevel = level
	}

	override fun getSaturationLevel(): Float =
		wrappedEntity.hungerManager.saturationLevel

	override fun setSaturationLevel( level: Float ) {
		wrappedEntity.hungerManager.saturationLevel = level
	}

	override fun getGamemode(): Gamemode {
		val mode = if (wrappedEntity.getWorld().isClient)
			MinecraftClient.getInstance().interactionManager!!.currentGameMode
		else
			(wrappedEntity as ServerPlayerEntity).interactionManager.gameMode
		if (mode == GameMode.DEFAULT)
			return Gamemode.SURVIVAL
		return Gamemode.valueOf(mode.getName().uppercase(Locale.getDefault()))
	}

	override fun getExperienceLevel(): Int =
		wrappedEntity.experienceLevel
}
